src/julia.jl

```
@@ -33,4 +33,56 @@ function multireshape(parent::AbstractArray, dimss::Array{Tuple{Vararg{Int}}})
     multireshape(parent, dimss...)
 end
 
+"""
+    pairwise(f::Function, a; kwargs...)
+
+Apply the function `f` to all pairwise combinations of elements from `a`. Set
+keyword `symmetric = true` to only calculate `f(x,y)` when `f(x,y) == f(y,x)`.
+"""
+function pairwise(f::Function, a; symmetric = false)
+    n = length(a)
+    first = f(a[1], a[1])
+    r = Matrix{typeof(first)}(n, n)
+    pairwise!(f, a, r; symmetric = symmetric)
+    r
+end
+
+"""
+    pairwise!(f::Function, a, r)
+
+Apply the function `f` to all pairwise combinations of elements from `a` and
+storing the results in `r`. Setting keyword `symmetric = true` will only calculate
+each variable combination once for symmetric functions. This is useful if `f` is
+expensive.
+"""
+function pairwise!(f::Function, a, r::AbstractMatrix; symmetric = false)
+    n = length(a)
+    size(r) == (n, n) || throw(DimensionMismatch("Incorrect size of r ($(size(r)), should be $((n, n)))"))
+    if symmetric
+        pairwise_symmetric!(f, a, r, n)
+    else
+        pairwise_default!(f, a, r, n)
+    end
+end
+
+function pairwise_default!(f, a, r, n)
+    @inbounds for j = 1:n
```

probably `for j in eachindex(a)` would be better, to handle other kinds of indices?  Then you couldn't take an `n` argument.

---

I'd still need to pass n to the `pairwise_symmetric` function, though, right, to enable `for i in j:n`? I think the `symmetric` keyword should not change what kind of objects (indexable vs iterable) the function should accept.

---

Objects with `eachindex` are still indexable.

You could do:
```
@inbounds for j in eachindex(a)
    i = j
    while i ≤ endof(a)
        r[i,j] = f(a[i],a[j])
        i = nextind(a, i)
    end
end
```
By using `nextind`, it will even work if `a` is a Unicode string, for pairwise functions over characters.

---

(This is one of those things where writing library code is much harder that writing functions in a particular program.  It's tricky to make a function like this extremely generic.)

---

But very cool to think about these things and get into the habit. I like the nextind approach a lot. I will experiment with it.

---

(However, you wouldn't want to do this in the case where the user called `pairwise!` ... in that case I would just throw an error.)

Of course, it may not be worthwhile to go to this trouble here.  For now, you could just do `r[i,j] = f(a[i],a[j])` and it will throw an error if the types are incompatible

---

I assume this last comment is to your suggested comment in the thread above about how to deal with type-unstable functions? I will keep the behaviour you suggest here - so that it errors if the types do not match.

---

Yes.